File: Windows_and_Linux/app_profile_service.py
"""
App Profile Service
-------------------

Singleton-style service for managing app-aware prompt profiles.
Persists profiles to app_profiles.json alongside config.json.
"""

import json
import logging
import os
import sys
from typing import List, Optional

from active_app import ActiveAppContext
from app_profile import AppMatcher, AppProfile, FormattingPreset


class AppProfileService:
    """Manages app-aware prompt profiles: CRUD, persistence, and prompt enrichment."""

    _instance: Optional["AppProfileService"] = None

    def __init__(self):
        self.profiles: List[AppProfile] = []
        self._file_path = os.path.join(os.path.dirname(sys.argv[0]), "app_profiles.json")
        self.load()

    @classmethod
    def shared(cls) -> "AppProfileService":
        if cls._instance is None:
            cls._instance = cls()
        return cls._instance

    # ── Persistence ──────────────────────────────────────────────

    def load(self):
        if os.path.exists(self._file_path):
            try:
                with open(self._file_path, "r", encoding="utf-8") as f:
                    data = json.load(f)
                self.profiles = [AppProfile.from_dict(d) for d in data]
                logging.debug(f"Loaded {len(self.profiles)} app profiles")
            except Exception as e:
                logging.error(f"Failed to load app profiles: {e}")
                self.profiles = []
        else:
            self.profiles = self._starter_profiles()
            self.save()

    def save(self):
        try:
            with open(self._file_path, "w", encoding="utf-8") as f:
                json.dump([p.to_dict() for p in self.profiles], f, indent=4)
            logging.debug(f"Saved {len(self.profiles)} app profiles")
        except Exception as e:
            logging.error(f"Failed to save app profiles: {e}")

    # ── CRUD ─────────────────────────────────────────────────────

    def add_profile(self, profile: AppProfile):
        self.profiles.append(profile)
        self.save()

    def update_profile(self, profile: AppProfile):
        for i, p in enumerate(self.profiles):
            if p.id == profile.id:
                self.profiles[i] = profile
                self.save()
                return
        # If not found, add it
        self.add_profile(profile)

    def delete_profile(self, profile_id: str):
        self.profiles = [p for p in self.profiles if p.id != profile_id]
        self.save()

    # ── Resolution ───────────────────────────────────────────────

    def resolve_profile(self, ctx: ActiveAppContext) -> Optional[AppProfile]:
        """Return the first enabled profile whose matchers match the context."""
        for profile in self.profiles:
            if profile.matches(ctx):
                return profile
        return None

    def enrich_system_instruction(self, instruction: str, ctx: ActiveAppContext) -> str:
        """
        Append profile-specific instructions to *instruction*.
        If no profile matches, append a plain-text fallback.
        """
        profile = self.resolve_profile(ctx)
        if profile:
            return f"{instruction}\n\n[App-Aware Formatting Instructions]\n{profile.combined_instructions}"
        # Invisible plain-text fallback
        return (
            f"{instruction}\n\n[Formatting Instructions]\n"
            "Format your response as plain text only. "
            "Do not use any Markdown syntax such as headers (#), bold (**), italic (*), "
            "bullet points (-), numbered lists, code blocks, or links."
        )

    # ── Starter profiles ─────────────────────────────────────────

    @staticmethod
    def _starter_profiles() -> List[AppProfile]:
        return [
            AppProfile(
                name="VS Code / Cursor",
                matchers=[
                    AppMatcher("app_name_contains", "Code"),
                    AppMatcher("app_name_contains", "Cursor"),
                ],
                formatting_preset=FormattingPreset.CODE_FRIENDLY,
            ),
            AppProfile(
                name="Obsidian",
                matchers=[AppMatcher("app_name_contains", "Obsidian")],
                formatting_preset=FormattingPreset.MARKDOWN,
            ),
            AppProfile(
                name="Slack",
                matchers=[AppMatcher("app_name_contains", "Slack")],
                formatting_preset=FormattingPreset.PLAIN_TEXT,
            ),
            AppProfile(
                name="Teams",
                matchers=[AppMatcher("app_name_contains", "Teams")],
                formatting_preset=FormattingPreset.PLAIN_TEXT,
            ),
            AppProfile(
                name="Email Clients",
                matchers=[
                    AppMatcher("app_name_contains", "Outlook"),
                    AppMatcher("app_name_contains", "Thunderbird"),
                    AppMatcher("app_name_contains", "Mail"),
                ],
                formatting_preset=FormattingPreset.EMAIL_FRIENDLY,
            ),
        ]

File: macOS/WritingTools/Services/AppProfileService.swift
import Foundation
import Observation
import AppKit

/// Manages app-aware prompt profiles and resolves the correct profile at command execution time.
@Observable
@MainActor
final class AppProfileService {

    static let shared = AppProfileService()

    // MARK: - State

    var profiles: [AppProfile] = [] {
        didSet { saveProfiles() }
    }

    private let saveKey = "app_profiles"
    private let hasInitializedKey = "has_initialized_app_profiles"

    // MARK: - Init

    private init() {
        loadProfiles()
    }

    // MARK: - Public API

    /// Returns the first enabled profile whose matchers fire for the given context,
    /// or nil if no profile matches.
    func resolveProfile(for ctx: ActiveAppContext) -> AppProfile? {
        profiles.first { $0.isEnabled && $0.matches(ctx) }
    }

    /// Appends formatting instructions to `basePrompt` based on the matched profile
    /// for `app`. Falls back to a plain-text hint when no profile matches.
    func enrichSystemPrompt(_ basePrompt: String, for app: NSRunningApplication?) -> String {
        guard let app else { return basePrompt }

        let ctx = ActiveAppContext(from: app)
        let instructions: String

        if let profile = resolveProfile(for: ctx) {
            instructions = profile.combinedInstructions
        } else {
            // Always-on plain-text fallback — safe for any unknown app.
            instructions = FormattingPreset.plainText.instructionText
        }

        return basePrompt + "\n\n" + instructions
    }

    // MARK: - CRUD

    func addProfile(_ profile: AppProfile) {
        profiles.append(profile)
    }

    func updateProfile(_ profile: AppProfile) {
        if let idx = profiles.firstIndex(where: { $0.id == profile.id }) {
            profiles[idx] = profile
        }
    }

    func deleteProfile(_ profile: AppProfile) {
        profiles.removeAll { $0.id == profile.id }
    }

    func moveProfiles(fromOffsets source: IndexSet, toOffset destination: Int) {
        profiles.move(fromOffsets: source, toOffset: destination)
    }

    // MARK: - Persistence

    private func loadProfiles() {
        let hasInitialized = UserDefaults.standard.bool(forKey: hasInitializedKey)

        if !hasInitialized {
            profiles = Self.starterProfiles
            saveProfiles()
            UserDefaults.standard.set(true, forKey: hasInitializedKey)
            return
        }

        if let data = UserDefaults.standard.data(forKey: saveKey),
           let decoded = try? JSONDecoder().decode([AppProfile].self, from: data) {
            profiles = decoded
        } else {
            profiles = Self.starterProfiles
        }
    }

    private func saveProfiles() {
        if let encoded = try? JSONEncoder().encode(profiles) {
            UserDefaults.standard.set(encoded, forKey: saveKey)
        }
    }

    // MARK: - Starter Profiles

    static let starterProfiles: [AppProfile] = [
        AppProfile(
            name: "Obsidian",
            matchers: [
                AppMatcher(bundleIdentifierEquals: "md.obsidian", appNameContains: "Obsidian")
            ],
            formattingPreset: .markdown,
            additionalInstructions: "You are writing into an Obsidian vault. Wikilinks ([[link]]) are acceptable."
        ),
        AppProfile(
            name: "VS Code / Cursor",
            matchers: [
                AppMatcher(bundleIdentifierEquals: "com.microsoft.VSCode", appNameContains: ""),
                AppMatcher(bundleIdentifierEquals: "com.todesktop.230313mzl4w4u92", appNameContains: "Cursor")
            ],
            formattingPreset: .codeFriendly,
            additionalInstructions: "You are editing inside a code editor. Preserve indentation and use fenced code blocks with the correct language tag."
        ),
        AppProfile(
            name: "Microsoft Teams",
            matchers: [
                AppMatcher(bundleIdentifierEquals: "com.microsoft.teams2", appNameContains: "Teams")
            ],
            formattingPreset: .plainText,
            additionalInstructions: "You are writing a Teams message. Keep it concise and conversational."
        ),
        AppProfile(
            name: "Slack",
            matchers: [
                AppMatcher(bundleIdentifierEquals: "com.tinyspeck.slackmacgap", appNameContains: "Slack")
            ],
            formattingPreset: .plainText,
            additionalInstructions: "You are writing a Slack message. Use lightweight formatting only (bold with *word*, code with `code`)."
        ),
        AppProfile(
            name: "Apple Mail",
            matchers: [
                AppMatcher(bundleIdentifierEquals: "com.apple.mail", appNameContains: "Mail")
            ],
            formattingPreset: .emailFriendly,
            additionalInstructions: "You are composing an email in Apple Mail. Use a professional tone and proper salutation/closing where appropriate."
        )
    ]
}
